doc/newmail.1
.if n \{\
.	ds ct "
.\}
.if t \{\
.	ds ct \\(co
.\}
.TH NEWMAIL 1L "Elm Version 2.5" "USENET Community Trust"
.SH NAME
newmail,wnewmail - programs to asynchronously notify of new mail
.SH SYNOPSIS
.B newmail
.B [-d]
.B [-i \fIinterval\fB]
.B [-w]
\fIfile-spec\fR {\fIfile-spec\fR...}
.br
.B wnewmail
.B [-d]
.B [-i \fIinterval\fB]
.B [-w]
\fIfile-spec\fR {\fIfile-spec\fR...}
.PP
.SH DESCRIPTION
.I Newmail\^
is a program to allow monitoring of mailboxes in an intelligent
fashion.  It is based loosely on \fIbiff(1)\fR and the version
of \fInewmail\fR that was distributed with \fBElm 1.7\fR.
.P
The basic operation is that the program will check the list of
specified mailboxes each \fIinterval\fR seconds and will list
any new mail that has arrived in any of the mailboxes,
indicating the sender name, and the subject of the message.
.P
Each entry displayed can be in a number of different formats
depending on the mode of the program, the number of folders
being monitored, and the status of the message.  If you're 
running it as a \fIwindow\fR (e.g. ``-w''
or invoked as \fIwnewmail\fR) then the output will be
similar to:
.nf

   \fIsender name\fR - \fIsubject of message\fR
   Priority: \fIsender name\fR - \fIsubject of message\fR

.fi
where <sender name> is either the name of the person sending it,
if available (the ARPA 'From:' line) or some other brief
indication of origin.   If you are the sender, <sender name> will be
replaced by "to <recipient name>".  If there
is no subject, the message "<no subject>" will appear on
the screen.
.P
Folders are indicated by having the folder name appear first
on the output line, as in:
.nf

   \fIfolder\fR: \fIsender name\fR - \fIsubject of message\fR

.fi
If you're running \fInewmail\fR without the windows option,
then the output is more suitable for popping up on an otherwise
active screen, and will be formatted:
.nf

   >> New mail from \fIsender name\fR - \fIsubject of message\fR
   >> Priority mail from \fIsender name\fR - \fIsubject of message\fR

.fi
Again, with folder names indicated as a prefix.
.P
The flags available are:
.TP 1.0i
.B "-d"
This will turn on the debugging, verbose output mode.  It is not
recommended that you use this option unless you're interested in
actually debugging the program.
.TP
.B "-i \fIinterval\fR  "
This will change the frequency that the program checks the folders
to the interval specified, in seconds.  The default interval for
the program is 60 seconds.  \fINote: if you change the interval
to less than 10 seconds, the program will warn you that it isn't
recommended.\fR
.TP
.B "-w"
Use of the ``-w'' flag will simulate having the program run
from within a window (e.g. the more succinct output format,
and so on).  Most likely, rather than using this option you
should be simply invoking \fIwnewmail\fR instead.
.P
\fIFile specs\fR are made up of two components, the
\fIfolder name\fR and the \fIprefix string\fR, the
latter of which can always be omitted.
The format is \fIfoldername=prefixstring\fR, and
you can specify folders by full name, by simply
the name of the user whose mailbox should be 
monitored, or by the standard \fBElm\fR 
metacharacters to specify your folder 
directory (e.g. ``+'', ``='', or ``%'').
.P
Folders that cannot be opened due to permission mismatches
will cause the program to immediately abort.  On the other
hand, files that do not exist will continue to be checked
every \fIinterval\fR seconds, so some care should be 
exercised when invoking the program.
.P
The program will run until you log out or explicitly kill
it, and can internally reset 
itself if any of the folders shrink in size and 
then grow again.
.P
The default folder to monitor is always your incoming mailbox.
.SH EXAMPLES
Some example invocations:
.nf

	$ \fBnewmail\fR

.fi
will check your incoming mailbox every 60 seconds.
.nf

	$ \fBnewmail  -i  15  joe  root\fR

.fi
will monitor the incoming mailboxes for ``joe'' and ``root'',
checking every 15 seconds for new mail.
.nf

	$ \fBnewmail  "mary=Mary"  +postmaster=POBOX\fR

.fi
will monitor the incoming mailbox for user ``mary'', prefixing
all messages with the string ``Mary'', and the folder in 
the users \fImaildir\fR directory called ``postmaster'', 
prefixing all of those messages with ``POBOX''.
.P
You can also have more complex monitoring too, for example:
.nf

	$ \fBnewmail  -i  30  $LOGNAME=me  ${LOGNAME}su=myroot   /tmp/mbox\fR

.fi
will check every 30 seconds for the users mailbox, a mailbox that
is the users login-name with ``su'' appended (e.g. ``joe'' would
become ``joesu'') and the file /tmp/mbox, prefixing new mail
for each with ``me'', ``myroot'' and ``mbox'' respectively.
.SH AUTHOR
Elm Development Group
.SH SEE ALSO
notify in sh(1) or csh(1)
.SH BUG REPORTS TO
Bill Pemberton  flash@virginia.edu
.SH COPYRIGHTS
\fB\*(ct\fRCopyright 1988-1995 by The USENET Community Trust
.br
Derived from Elm 2.0, \fB\*(ct\fR Copyright 1986, 1987 by Dave Taylor


Changes
Changes in 2.5 PL8

- fixed a hang when hitting certain 8-bit chars in the From line

- fixed a file descriptor leak (from Andreas Luik)

- removed all Expires: processing


Changes in 2.5 PL7

- a few display issues fixed (from Kenneth Reek)

- buffer overflow in frm fixed

Changes in 2.5 PL6

- Fixed message-id in the In-Reply-To header.

- Actually show the -r option in the usage

- Whacked a few more places where we might follow a symlink in tmpdir

- Fastmail didn't grok commas correctly.

- Change lock.c to not use errno unless there's an errno

Changes in 2.5 PL5

- A few more buffer overruns fixed.

Changes in 2.5 PL4

- little bug in aliasdb.c (from Scot W. Hetzel <scot@genroco.com>)

- Message-Id: could overflow (from Darin Wayrynen <darin@deru.net>)

- Fixed the permissions setting problem with the edit mailbox function

- Added a ResizeScreen() call after show_msg() calls to hopefully deal wit
  SIGWINCH problems (inspired from Yuke Zhuge <zhuge@stone.pdx.informix.com>)
  
- Fixed problems under SunOS 4.1.3 (from Paul J. Menchini <mench@mench.com>)

- Fixed content encoding (from David Woolley  <david@djwhome.demon.co.uk>)

- Attempt to stop crashes with long recipient list (from Gilles Detillieux 
  <grdetil@scrc.umanitoba.ca>)
  

Changes in 2.5 PL3

- Fixed the declaration of outbound_from_name (from Richard Jackson 
  <rjackson@gmu.edu>)

- Fixed the leapyear macro (from Gerhard Niklasch)
  
- Fixed editing of a long CC line (from Derek Godfrey <djg@sjc.ascend.com>)

- Mail that lands in the sentmail with attachments now has the 
  Content-disposition header text added to the message (from Dave 
  Spensley)
  
- Made allto a bit longer (suggested by Dave Spensley 
  <davesp@typhoon.xnet.com>)

- Fixed Configure to find (again) fcntl on Linux

- can_access will now allow you to write to a char device (/dev/null)

Changes in 2.5 PL2

- Fixed (again, maybe?) savebyalias from Adam Benjamin <aebenjam@opentext.com>

- Fixed bug were an alias delete would crash elm if the mailbox was empty

Changes in 2.5 PL1

- Fixed savebyalias from Michael D Hughes <mdh@logcabin.woods.bridge.com> 
  and Robert Schwartz <rjs@elt.com>
  
- Changes for 8 bit clean from H.J. Lu <hjl@varesearch.com>

Changes in 2.5 PL0

- One minor change to make an easier compile on SparcLinux.

Changes in 2.5 PL0pre8

- Y2K stuff from Ed Symanzik <zik@msu.edu>

Changes in 2.5 PL0pre7

- minor change to deal with egcs

Changes in 2.5 PL0pre6

- newalias nows does a geteuid() instead of getuid().  This is to lets 
  SysAdmins do "su <user> -c 'newalias'"
  

Changes in 2.5 PL0pre5

- Fix for GECOS entries with comma-delimited sub-fields from Chip 
  Rosenthal <chip@unicom.com>.

- Fix? for keys getting screw up after a ^Z from Chip Rosenthal
  <chip@unicom.com>.


Changes in 2.5 PL0pre4

- A few changes for BSD from Chip Rosenthal <chip@unicom.com>.

- A few changes to Configure suggested by Chip Rosenthal <chip@unicom.com>.


Changes in 2.5 PL0pre3

- Configure didn't see fcntl locking on Linux.  Fix from Sean McDermott 
  <spm@intel.com>


Changes in 2.5 beta 3

- Fixed a problem where elm would try to close a file twice when
  sending an attachment (from reichler@ews.uiuc.edu)

- Added docs for environment variables to the man page (from Eric
  Haszlakiewicz <nimenees@uaccess.net>)

Changes in 2.5 beta 2

- IncomingFolders no longer set Magic Mode on for everything

- Configure should now deal with Linux.

- Make //pattern honor current limit (from Keith Neufeld<neufeld@pvi.org>)

- year 2000 fix.....?

- Building on Solaris requires /usr/ccs/bin to be in the PATH

Changes in 2.5 beta 1

- b)ounce honor DONT_ADD_FROM (from Keith Neufeld<neufeld@pvi.org>)

- removed [encode] display and reply (from Keith Neufeld<neufeld@pvi.org>)

- make right/left arrows work in the file browser the same as in the
  message index. (from Keith Neufeld<neufeld@pvi.org>)

- fixed a bug in the incomingfolders code (it wouldn't save) (from Bob Schwartz
  <rjs@elt.com>)
  
- fixed the meaning of noheader for forwards (from Keith Neufeld<neufeld@pvi.org>)



Changes in 2.5 alpha 11

- fixed a bug in the alias code (thanks to Keith Neufeld<neufeld@pvi.org>)

- fixed the generation of temp file names in newmbox.c

- allowed longer Cc: lines.  This was due to a crash with a group reply.

- added incomingfolders to the elmrc.  This is a list of folders that will
  automatically be opened with "magic mode" on. (inspired by a patch from
  Cristophe Kalt <kalt@ensta.fr>)


Changes in 2.5 alpha 10

- Press u on a message that is read but not deleted will mark the message as
  unread (from Larry Schwimmer <rosebud@cyclone.Stanford.EDU>).

- mmencode.1 man page is now installed with others (from Gregory Margo
  <gmargo@newton.vip.best.com>).

- Magic mode stuff is now in help and NLS (from Gregory Margo
  <gmargo@newton.vip.best.com>).

- allow elimination of aliases from larger aliases (To: bigalias -subalias).

utils/newmail.c


/*******************************************************************************
 *  The Elm Mail System  -  $Revision: 1.7 $   $State: Exp $
 *
 *                      Copyright (c) 1988-1995 USENET Community Trust
 * 			Copyright (c) 1986,1987 Dave Taylor
 *******************************************************************************
 * Bug reports, patches, comments, suggestions should be sent to:
 *
 *      Bill Pemberton, Elm Coordinator
 *      flash@virginia.edu
 *
 *******************************************************************************
 * $Log: newmail.c,v $
 * Revision 1.7  1996/03/14  17:30:10  wfp5p
 * Alpha 9
 *
 * Revision 1.6  1995/09/29  17:42:46  wfp5p
 * Alpha 8 (Chip's big changes)
 *
 * Revision 1.5  1995/07/18  19:00:13  wfp5p
 * Alpha 6
 *
 * Revision 1.4  1995/06/14  19:59:42  wfp5p
 * Changes for alpha 3.
 * Changed the #ifdef for defining UTIMBUF
 *
 * Revision 1.3  1995/05/30  18:27:31  wfp5p
 * Added fix for utime() from Chip Rosenthal <chip@unicom.com> .
 *
 * Revision 1.2  1995/04/20  21:02:09  wfp5p
 * Added the showreply feature and emacs key bindings.
 *
 * Revision 1.1.1.1  1995/04/19  20:38:41  wfp5p
 * Initial import of elm 2.4 PL0 as base for elm 2.5.
 *
 ******************************************************************************/

/** This is actually two programs folded into one - 'newmail()' and
    'wnewmail()'.  They perform essentially the same function, to
    monitor the mail arriving in a set of/a mailbox or folder, but
    newmail is designed to run in background on a terminal, and
    wnewmail is designed to have a window of its own to run in.

    The main difference is that wnewmail checks for mail more often.

    The usage parameters are:

	-i <interval>  		how often to check for mail
				(default: 60 secs if newmail,
					  10 secs if wnewmail)

	<filename>		name of a folder to monitor
				(can prefix with '+'/'=', or can
			 	default to the incoming mailbox)

	<filename>=prefix	file to monitor, output with specified
				prefix when mail arrives.

    If we're monitoring more than one mailbox the program will prefix
    each line output (if 'newmail') or each cluster of mail (if 'wnewmail')
    with the basename of the folder the mail has arrived in.  In the 
    interest of exhaustive functionality, you can also use the "=prefix"
    suffix (eh?) to specify your own strings to prefix messages with.

    The output format is either:

	  newmail:
	     >> New mail from <user> - <subject>
	     >> Priority mail from <user> - <subject>

	     >> <folder>: from <user> - <subject>
	     >> <folder>: Priority from <user> - <subject>

	  wnewmail:
	     <user> - <subject>
	     Priority: <user> - <subject>

	     <folder>: <user> - <subject>
	     <folder>: Priority: <user> - <subject>\fR

**/

#define INTERN
#include "elm_defs.h"
#include "s_newmail.h"
#include "port_stat.h"

#ifdef I_TIME
#  include <time.h>
#endif
#ifdef I_SYSTIME
#  include <sys/time.h>
#endif
#ifdef BSD
#  include <sys/timeb.h>
#endif

#ifdef I_UTIME
#  include <utime.h>
#endif
#ifdef I_SYSUTIME
#  include <sys/utime.h>
#endif


/**********
   Since a number of machines don't seem to bother to define the utimbuf
   structure for some *very* obscure reason.... 

   Suprise, though, BSD has a different utime() entirely...*sigh*
**********/

/*if defined (BSD) && !defined(UTIMBUF) */
#if !defined(UTIMBUF)
struct utimbuf {
        time_t  actime;         /** access time       **/ 
        time_t  modtime;        /** modification time **/
       };

#endif


#define LINEFEED		(char) 10
#define BEGINNING		0			/* seek fseek(3S) */
#define DEFAULT_INTERVAL	60

#define MAX_FOLDERS		25		/* max we can keep track of */

/*
 * The "read_headers()" and "show_header()" routines use a (struct header_rec)
 * to hold the header information.  This structure does not have a flag to
 * save the message priority status.  We don't need the "encrypted" flag in
 * the structure, so we will use that instead.
 */
#define priority encrypted

#define metachar(c)	(c == '+' || c == '=' || c == '%')

long  bytes();

struct folder_struct {
	  char		foldername[SLEN];
	  char		prefix[NLEN];
	  long		filesize;
	  int		access_error;
       } folders[MAX_FOLDERS] = {0};

int  interval_time,		/* how long to sleep between checks */
     debug = 0,			/* include verbose debug output?    */
     in_window = 0,		/* are we running as 'wnewmail'?    */
     total_folders = 0,		/* # of folders we're monitoring    */
     print_prefix = 0,		/* force printing of prefix	    */
     current_folder = 0;	/* struct pointer for looping       */
FILE	*fd = NULL;		/* fd to use to read folders	    */

#ifdef PIDCHECK
int  parent_pid;		/* See if sucide should be attempt  */
#endif /* PIDCHECK */

extern int errno;

#if defined(BSD) && !defined(UTIMBUF)
        time_t utime_buffer[2];         /* utime command */
#else
        struct utimbuf utime_buffer;    /* utime command */
#endif

static char	*no_subj,	/* Pointer to No subject text	*/
		*priority_to,	/* pointer to Priority to text	*/
		*priority_text,	/* pointer to priority text	*/
		*To_text,	/* pointer To to text		*/
		*priority_mail,	/* pointer to priority mail	*/
		*mail_text,	/* pointer to mail text		*/
		*to_text,	/* pointer to to text		*/
		*from_text;	/* pointer to from text		*/

main(argc, argv)
int argc;
char *argv[];
{
	extern char *optarg;
	extern int   optind;
	char *ptr;
	int c, i, done;
	long lastsize,
	     newsize;			/* file size for comparison..      */
	register struct folder_struct *cur_folder;

	initialize_common();

	/* Get the No subject string */

	no_subj = catgets(elm_msg_cat, NewmailSet, NewmailNoSubject,
	   "(No Subject Specified)");
	priority_to = catgets(elm_msg_cat, NewmailSet,
	   NewmailInWinPriorityTo, "Priority to ");
	priority_text = catgets(elm_msg_cat, NewmailSet,
	      NewmailInWinPriority, "Priority ");
	To_text = catgets(elm_msg_cat, NewmailSet, NewmailInWinTo, "To ");
	priority_mail = catgets(elm_msg_cat, NewmailSet,
	   NewmailPriorityMail, "Priority mail ");
	mail_text = catgets(elm_msg_cat, NewmailSet, NewmailMail, "Mail ");
	to_text = catgets(elm_msg_cat, NewmailSet, NewmailTo, "to ");
	from_text = catgets(elm_msg_cat, NewmailSet, NewmailFrom, "from ");

#ifdef PIDCHECK				/* This will get the pid that         */
	parent_pid = getppid();		/* started the program, ie: /bin/sh   */
					/* If it dies for some reason (logout)*/
#endif /* PIDCHECK */			/* Then exit the program if PIDCHECK  */

	interval_time = DEFAULT_INTERVAL;

	/** let's see if the first character of the basename of the
	    command invoked is a 'w' (e.g. have we been called as
	    'wnewmail' rather than just 'newmail'?)
	**/

	for (i=0, ptr=(argv[0] + strlen(argv[0])-1); !i && ptr > argv[0]; ptr--)
	  if (*ptr == '/') {
	    in_window = (*(ptr+1) == 'w');
	    i++;
	  }

	if (ptr == argv[0] && i == 0 && argv[0][0] == 'w')
	  in_window = 1;

	while ((c = getopt(argc, argv, "di:w")) != EOF) {
	  switch (c) {
	    case 'd' : debug++;					break;
	    case 'i' : interval_time = atoi(optarg);		break;
	    case 'w' : in_window = 1;				break;
	    default  : usage(argv[0]);				exit(1);
	 }
	}

	if (interval_time < 10)
	  if (interval_time == 1)
	    fprintf(stderr, catgets(elm_msg_cat, NewmailSet, NewmailShort,
"Warning: interval set to 1 second.  I hope you know what you're doing!\n"));
	  else
	    fprintf(stderr, catgets(elm_msg_cat, NewmailSet, NewmailShortPlur,
"Warning: interval set to %d seconds.  I hope you know what you're doing!\n"),
		interval_time);

	/* now let's parse the foldernames, if any are given */

	if (optind >= argc) /* get default */
	  add_default_folder();
	else {
	  while (optind < argc)
	    add_folder(argv[optind++]);
	  pad_prefixes();			/* for nice output...*/
	}
	if (total_folders > 1)
		print_prefix = 1;

#ifdef AUTO_BACKGROUND
	if (! in_window) {
	  if (fork())	    /* automatically puts this task in background! */
	    exit(0);

	  (void) signal(SIGINT, SIG_IGN);
	  (void) signal(SIGQUIT, SIG_IGN);
	}
#endif
	(void) signal(SIGHUP, SIG_DFL);

	if (in_window && ! debug)
	  printf(catgets(elm_msg_cat, NewmailSet, NewmailIncommingMail,
	      "Incoming mail:\n"));

	while (1) {

#ifdef PIDCHECK
	if ( kill(parent_pid,0))
		exit(0);
#else
#ifndef AUTO_BACKGROUND		/* won't work if we're nested this deep! */
	  if (getppid() == 1) 	/* we've lost our shell! */
	    exit(0);
#endif /* AUTO_BACKGROUND */
#endif /* PIDCHECK */

	  if (! isatty(1))	/* we're not sending output to a tty any more */
	     exit(0);

	  dprint(1, (debugfile, "\n----\n"));

	  for (i = 0; i < total_folders; i++) {

	    cur_folder = &folders[i];
	    dprint(1, (debugfile, "[checking folder #%d: %s]\n",
		i, cur_folder->foldername));

	    if ((newsize = bytes(cur_folder->foldername)) == 
	        cur_folder->filesize) 	/* no new mail has arrived! */
	    	continue;

	    if ((fd = fopen(cur_folder->foldername,"r")) == NULL) {
	      if (errno == EACCES) {
		cur_folder->access_error++;
		if (cur_folder->access_error > 5) {
		  fprintf(stderr, catgets(elm_msg_cat, NewmailSet,
		      NewmailErrNoPerm,
		      "\nPermission to monitor \"%s\" denied!\n\n"),
			 cur_folder->foldername);
		  sleep(5);
		  exit(1);
		}
	      }
	      continue;
	    }

	    if ((newsize = bytes(cur_folder->foldername)) > 
	        cur_folder->filesize) {	/* new mail has arrived! */

	      dprint(1, (debugfile,
		  "\tnew mail has arrived!  old size = %ld, new size=%ld\n",
		  cur_folder->filesize, newsize));

	      /* skip what we've read already... */

	      if (fseek(fd, cur_folder->filesize, 
			BEGINNING) != 0)
	        perror("fseek()");

	      cur_folder->filesize = newsize;

	      /* read and display new mail! */
	      if (read_headers(cur_folder) && ! in_window)
	        printf("\n\r");
	      /* try to set the file access times back, ignore
		 failures */

#if defined(BSD) && !defined(UTIMBUF)
	      utime(cur_folder->foldername, utime_buffer);
#else
              utime(cur_folder->foldername, &utime_buffer);
#endif
	    }
	    else {	/* file SHRUNK! */

	      cur_folder->filesize = bytes(cur_folder->foldername);
	      lastsize = cur_folder->filesize;
	      done     = 0;

	      while (! done) {
	        sleep(1);	/* wait for the size to stabilize */
	        newsize = bytes(cur_folder->foldername);
	        if (newsize != lastsize)
	          lastsize = newsize;
		else
	          done++;
	      } 
	        
	      cur_folder->filesize = newsize;
	    }
	    (void) fclose(fd);			/* close it and ...         */
	  }

	  sleep(interval_time);
	}
}

int
read_headers(cur_folder)
register struct folder_struct *cur_folder;
{
	/** read the headers, output as found given current_folder,
	    the prefix of that folder, and whether we're in a window
	    or not.
	**/

	struct header_rec hdr;		/* holds header info on curr mssg */
	char buffer[SLEN];		/* message line buffer		  */
	char *fld_name, *fld_val;	/* field name and value pointers  */
	register int in_header;		/* TRUE when processing a hdr	  */
	int count;			/* count of messages done	  */
	int init_header;		/* TRUE to reset hdr for new mssg */
	count = 0;
	init_header = TRUE;

	/*
	 * Go through every line of the mailbox.
	 */
	while (mail_gets(buffer, SLEN, fd) != 0) {

	  /*
	   * Initialize header record for a new message.
	   */
	  if (init_header) {
	    hdr.to[0] = '\0';
	    hdr.from[0] = '\0';
	    hdr.subject[0] = '\0';
	    hdr.priority = FALSE;
	    in_header = FALSE;
	    init_header = FALSE;
	  }

	  /*
	   * Search for the start of a message.
	   */
	  if (!in_header) {
	    in_header = 
	      strbegConst(buffer, "From ") && real_from(buffer, &hdr);
	    continue;
	  }

	  /*
	   * Look for blank line at the end of the message header.
	   */
	  if (buffer[0] == LINEFEED) {
	    count++;
	    show_header(&hdr, cur_folder);
	    init_header = TRUE;
	    continue;
	  }


	  if (header_cmp(buffer, "From", (char *) NULL))  {
	    parse_arpa_who(buffer+5, hdr.from);
	    continue;
	  }

	  /*
	   * Split up the header into a field name and field value.
	   */
	  fld_name = fld_val = buffer;
	  while (*fld_val != '\0' && !isspace(*fld_val))
	    ++fld_val;
	  if (*fld_val != '\0') {
	    *fld_val = '\0';
	    do {
	      ++fld_val;
	    } while (isspace(*fld_val));
	  }

	  if (istrcmp(fld_name, ">From") == 0) {
	    forwarded(fld_val, hdr.from);
	    continue;
	  }

	  if (istrcmp(fld_name, "To:") == 0) {
	    figure_out_addressee(fld_val, user_name, hdr.to);
	    continue;
	  }

	  if (istrcmp(fld_name, "Subject:") == 0) {
	    strfcpy(hdr.subject, fld_val, sizeof(hdr.subject));
	    continue;
	  }

	  if (istrcmp(fld_name, "Re:") == 0) {
	    if (hdr.subject[0] == '\0')
	      strfcpy(hdr.subject, fld_val, sizeof(hdr.subject));
	    continue;
	  }

	  if (istrcmp(fld_name, "Importance:") == 0) {
	    if (atoi(fld_val) >= 2)
	      hdr.priority = TRUE;
	    continue;
	  }

	  if (istrcmp(fld_name, "Priority:") == 0) {
	    if (strincmp(fld_val, "normal", 6) != 0 &&
	    		strincmp(fld_val, "non-urgent", 10) != 0)
	      hdr.priority = TRUE;
	    continue;
	  }

	  /*
	   * If we reach this point, it must be an unknown
	   * or boring header.  Skip it.
	   */

	}

	return(count);
}

add_folder(name)
char *name;
{
	/* add the specified folder to the list of folders...ignore any
	   problems we may having finding it (user could be monitoring
	   a mailbox that doesn't currently exist, for example)
	*/

	char *cp, buf[SLEN];

	if (current_folder >= MAX_FOLDERS) {
	  fprintf(stderr, catgets(elm_msg_cat, NewmailSet, NewmailErrMaxFolders,
            "Sorry, but I can only keep track of %d folders.\n"), MAX_FOLDERS);
	  exit(1);
	}

	/* now let's rip off the suffix "=<string>" if it's there... */

	for (cp = name + strlen(name); cp > name+1 && *cp != '=' ; cp--)
	  /* just keep stepping backwards */ ;

	/* if *cp isn't pointing to the first character we'e got something! */

	if (cp > name+1) {

	  *cp++ = '\0';		/* null terminate the filename & get prefix */

	  if (metachar(*cp)) cp++;

	  strcpy(folders[current_folder].prefix, cp);
	  print_prefix = 1;
	}
	else {			/* nope, let's get the basename of the file */
	  for (cp = name + strlen(name); cp > name && *cp != '/'; cp--)
	    /* backing up a bit... */ ;

	  if (metachar(*cp)) cp++;
	  if (*cp == '/') cp++;

	  strcpy(folders[current_folder].prefix, cp);
	}

	/* and next let's see what kind of weird prefix chars this user
	   might be testing us with.  We can have '+'|'='|'%' to expand
	   or a file located in the incoming mail dir...
	*/

	if (metachar(name[0]))
	  expand_filename(name, folders[current_folder].foldername);
	else if (access(name, 00) == -1) {
	  /* let's try it in the mail home directory */
	  sprintf(buf, "%s%s", mailhome, name);
	  if (access(buf, 00) != -1) 		/* aha! */
	    strcpy(folders[current_folder].foldername, buf);
	  else
	    strcpy(folders[current_folder].foldername, name);
	}
	else
	  strcpy(folders[current_folder].foldername, name);

	/* now let's try to actually open the file descriptor and grab
	   a size... */

	if ((fd = fopen(folders[current_folder].foldername, "r")) == NULL)
          if (errno == EACCES) {
	    fprintf(stderr, catgets(elm_msg_cat, NewmailSet, NewmailErrNoPerm,
		      "\nPermission to monitor \"%s\" denied!\n\n"),
			 folders[current_folder].foldername);
	    exit(1);
	  }

	folders[current_folder].filesize = 
	      bytes(folders[current_folder].foldername);

	/* and finally let's output what we did */

	  dprint(1, (debugfile, "folder %d: \"%s\" <%s> %s, size = %ld\n",
	      current_folder,
	      folders[current_folder].foldername,
	      folders[current_folder].prefix,
	      fd == NULL? "not found" : "opened",
	      folders[current_folder].filesize));

	if (fd != NULL) /* Close it only if we succeeded in opening it */
		(void) fclose(fd);

	/* and increment current-folder please! */

	current_folder++;
	total_folders++;
}

add_default_folder()
{

	/* this routine will add the users home mailbox as the folder
	 * to monitor.  Since there'll only be one folder we'll never
	 * prefix it either...
	 */
	strcpy(folders[0].foldername, incoming_folder);
	
	fd = fopen(folders[0].foldername, "r");
	folders[0].filesize = bytes(folders[0].foldername);

	dprint(1, (debugfile, "default folder: \"%s\" <%s> %s, size = %ld\n",
	      folders[0].foldername,
	      folders[0].prefix,
	      fd == NULL? "not found" : "opened",
	      folders[0].filesize));

	total_folders = 1;
	if (fd != NULL) /* Close it only if we succeeded in opening it */
	  fclose(fd);
}


forwarded(buffer, who)
char *buffer, *who;
{
	/** change 'from' and date fields to reflect the ORIGINATOR of 
	    the message by iteratively parsing the >From fields...
	    The leading >From will already be stripped off the line. **/

	char machine[SLEN], buff[SLEN];

	machine[0] = '\0';
	sscanf(buffer, "%s %*s %*s %*s %*s %*s %*s %*s %*s %s",
	            who, machine);

	if(machine[0] == '\0')	/* try for address with timezone in date */
	sscanf(buffer, "%s %*s %*s %*s %*s %*s %*s %*s %s",
	            who, machine);

	if (machine[0] == '\0') /* try for srm address */
	  sscanf(buffer, "%s %*s %*s %*s %*s %*s %*s %s",
	            who, machine);

	if (machine[0] == '\0')
	  sprintf(buff,"anonymous");
	else
	  sprintf(buff,"%s!%s", machine, who);

	strncpy(who, buff, SLEN);
}

show_header(hdr, cur_folder)
struct header_rec *hdr;
struct folder_struct *cur_folder;
{
	char from_line[SLEN];
	char prefix[SLEN];
	int used_to_line;

	if (hdr->from[0] == '\0')
	  strcpy(hdr->from, user_name);
	if (hdr->subject[0] == '\0')
	  strcpy(hdr->subject, no_subj);
	else {
	  used_to_line = strlen(hdr->subject) - 1;
	  if (hdr->subject[used_to_line] == '\n')
	    hdr->subject[used_to_line] = '\0'; /* strip off trailing new line, we add our own */
	}

	used_to_line = tail_of(hdr->from, from_line, hdr->to);
	prefix[0] = '\0';

	if (! in_window)
	  strcat(prefix, ">> ");

	if (print_prefix) {
	  strcat(prefix, cur_folder->prefix);
	  strcat(prefix, ": ");
	}

	if (in_window) {
	  if (hdr->priority && used_to_line)
	    strcat(prefix, priority_to);
	  else if (hdr->priority)
	    strcat(prefix, priority_text);
	  else if (used_to_line)
	    strcat(prefix, To_text);
	  printf("\007%s%s -- %s\n", prefix, from_line, hdr->subject);
	} else {
	  if (hdr->priority)
	    strcat(prefix, priority_mail);
	  else
	    strcat(prefix, mail_text);
	  if (used_to_line)
	    strcat(prefix, to_text);
	  else
	    strcat(prefix, from_text);
	  printf("\n\r%s%s - %s\n\r", prefix, from_line, hdr->subject);
	}
}

long
bytes(name)
char *name;
{
	/** return the number of bytes in the specified file.  This
	    is to check to see if new mail has arrived....  **/

	int ok = 1;
	extern int errno;	/* system error number! */
	struct stat buffer;

	if (stat(name, &buffer) != 0)
	  if (errno != 2) {
	    MCfprintf(stderr, catgets(elm_msg_cat, NewmailSet, NewmailErrFstat,
	      "Error %d attempting fstat on %s"), errno, name);
	    exit(1);
	  }
	  else
	    ok = 0;
	
	/* retain the access times for later use */

#if defined(BSD) && !defined(UTIMBUF)
        utime_buffer[0]     = buffer.st_atime;
        utime_buffer[1]     = buffer.st_mtime;
#else
        utime_buffer.actime = buffer.st_atime;
        utime_buffer.modtime= buffer.st_mtime;
#endif

	return(ok ? buffer.st_size : 0);
}


usage(name)
char *name;
{
	/* print a nice friendly usage message */

	fprintf(stderr, catgets(elm_msg_cat, NewmailSet, NewmailArgsHelp1,
"\nUsage: %s [-d] [-i interval] [-w] {folders}\n\
\targ\t\t\tMeaning\n\r\
\t -d  \tturns on debugging output\n\
\t -i D\tsets the interval checking time to 'D' seconds\n\
\t -w  \tforces 'window'-style output, and bypasses auto-background\n\n"),
	name);

	fprintf(stderr, catgets(elm_msg_cat, NewmailSet, NewmailArgsHelp2,
"folders can be specified by relative or absolute path names, can be the name\n\
of a mailbox in the incoming mail directory to check, or can have one of the\n\
standard Elm mail directory prefix chars (e.g. '+', '=' or '%%').\n\
Furthermore, any folder can have '=string' as a suffix to indicate a folder\n\
identifier other than the basename of the file\n\n"));

}


expand_filename(name, store_space)
char *name, *store_space;
{
	strcpy(store_space, name);
	if (expand(store_space) == 0) {
	  fprintf(stderr, catgets(elm_msg_cat, NewmailSet, NewmailErrExpand,
	    "Sorry, but I couldn't expand \"%s\"\n"),name);
	  exit(1);
	}
}

pad_prefixes()
{
	/** This simple routine is to ensure that we have a nice
	    output format.  What it does is whip through the different
	    prefix strings we've been given, figures out the maximum
	    length, then space pads the other prefixes to match.
	**/

	register int i, j, len = 0;

	for (i=0; i < total_folders; i++)
	  if (len < (j=strlen(folders[i].prefix)))
	    len = j;
	
	for (i=0; i < total_folders; i++)
	  for (j = strlen(folders[i].prefix); j < len; j++)
	    strcat(folders[i].prefix, " ");
}
